Add acceptance tests for `gh auth` commands

jtmcg opened the pull request:
## Changes

Adds acceptance tests for some of the `gh auth` commands. Some of them are very straight forward while others require a bit more setup to understand. I'll do my best to add that context here:

### `gh auth status` and `gh auth token` 

Work as expected without any additional setup

### `gh auth setup-git`

To test this, I decided to look into the .gitconfig used for the test and examine the credential helpers. However, the format of the git command is 

```sh
git config --get credential.<URL>.helper
```
What's awkward about this is that the `<URL>` depends on the host the user specified when running the tests, meaning I'd need to create a key like `credential.https://github.com.helper` to access what I need while setting this up. 

There was no functionality for string formatting before, so I added the command `formattedStringToEnv` which essentially wraps `fmt.Sprintf()` and saves the string to an environment variable.

This allowed me to dynamically create the config key in the test.

### `gh auth login`

The auth used in all the tests up until this point has used the `GH_TOKEN` env var to override the need to log in. Since we want to test logging in with this command directly, I've had to delete the `GH_TOKEN` env var but still save the token value so I can auth using the `--with-token` flag available with `gh auth login`.

Notably, the UX for the `gh auth login --with-token` command is pretty bad, but for now it works by specifying the token value to `stdin` just before running the `gh auth login --with-token` command. There is an [issue](https://github.com/cli/cli/issues/9789) to address this bad UX.

### `gh auth switch`

This test has been removed for two reasons:
1. It wasn't providing much value, as `gh auth switch` is a local command (doesn't interact with the API)
2. It introduced a good deal of complexity to the acceptance tests suite

The work for `gh auth switch` is now captured in this branch: [jtmcg/testscripts-auth-switch](https://github.com/cli/cli/tree/jtmcg/testscripts-auth-switch)

~~This is the trickiest command of them all, requiring us to log in using 2 sets of credentials. Thus, I've implemented some optional env vars that can be specified when a command is run to make this one work: `GH_ACCEPTANCE_USER`, `GH_ACCEPTANCE_HOST_ALT`, `GH_ACCEPTANCE_USER_ALT`, and `GH_ACCEPTANCE_TOKEN_ALT`, and `GH_ACCEPTANCE_TOKEN_ALT`. All of the `ALT` variables correspond to a second user account you are currently logged in to for testing.~~

~~With these additional variables, the test logs in to both instances and then tests that `gh auth switch` works to go between the two. I think the test is rather straight forward if the setup is not.~~

## Testing

Directions to run the testscripts are in the README. Here's the command to run just the auth tests, for convenience:

```
GH_ACCEPTANCE_TOKEN=<TOKEN> \
GH_ACCEPTANCE_ORG=<ORG> \
GH_ACCEPTANCE_HOST=<HOST> \
go test -tags=acceptance -run ^TestAuth$ ./acceptance
```

jtmcg committed:
Setup acceptance testing for auth and tests for auth-token and auth-status

jtmcg committed:
Add --token flag to `gh auth login` to accept a PAT as a flag

Additionally, this commit adds acceptance testing for `gh auth login` and
`gh auth logout`.

The --token flag was necessary for adding testing for `gh auth login`
because the current implementation with `--with-token` appears to be
broken. It hangs, waiting for user input, but user input doesn't exit it.

Additionally, it appears that `--with-token` is intended to allow for TTY
input of an auth token, but it isn't implemented.

`--with-token` does work when used with the redirect operator `<` when the
token is saved in a file. However, due to limitations of testscripts, I
could not use a file for saving the token in a repeatable manner. Thus,
implementing the `--token` flag seemed like a quick solution to validate
that the direction I was going during testing was valid.

Whether the flag stays or not is up for discussion, and I'd love to get
input on that from the team.

williammartin commented on acceptance/testdata/auth/auth-login-logout.txtar:
```diff
@@ -0,0 +1,29 @@
+# overrides are happening
+
+# Copy $GH_TOKEN to a new env var
+setEnvVar LOGIN_TOKEN $GH_TOKEN
```
I'll need to check this but I'm surprised that

```
env LOGIN_TOKEN $GH_TOKEN
```

Wouldn't work here. Did you check that?

jtmcg replied:
I'm getting this error when I try to use the code snippet above

I'm not sure why it isn't working, either

jtmcg replied:
It looks like there may be an extra `=` at the end of `LOGIN_TOKEN`, based on the output, but I'd still expect that to match on `GH_TOKEN` because we're using regex... Also, notably, it hangs when I comment out the assertions on lines 7 & 8, so something weird is happening

williammartin replied:
I think it needs to be `env LOGIN_TOKEN=$GH_TOKEN` sorry.

jtmcg replied:
Ah, that did it 👍 Good callout

jtmcg replied:
Now I can remove the two new shared functions 🥳

philipjonsen replied:
Hi, thanks for all the help first and foremost. I just realized that ethereum has jumped in and helped. I just saw open oracle and avalanche with a 1000+ security problems, I didn't know if they also are jumping on and working on this or who are. Just need a place to communicate better perhaps. also changed permission for some respitores to be extra secure. Now that I know that eth are working with us and hopefully microsoft we can fix eventuelly mistakes or security issues.

jtmcg committed:
Clean up auth-login-logout acceptance test with native functionality

The previous commit introduced two new functions, setEnvVar and
deleteEnvVar that are duplicative of functionality native to testscripts.
This commit switches to the native functionality and removes the
duplicative functions introduced in the previous commit.

Additionally, it removes the `--token` flag that was added to
`gh auth login`

jtmcg committed:
Add acceptance tests for auth-setup-git and formattedStringToEnv helper func

To test this, I decided to look into the .gitconfig used for the test and
examine the credential helpers. However, the format of the git command is

`git config --get credential.<URL>.helper`

What's awkward about this is that the <URL> depends on the host the user
specified when running the tests, meaning I'd need to create a key like
credential.https://github.com.helper to access what I need while setting
this up.

There was no functionality for string formatting before, so I added the
command formattedStringToEnv which essentially wraps fmt.Sprintf() and
saves the string to an environment variable.

This allowed me to dynamically create the config key in the test.

jtmcg committed:
Remove comment from gh auth logout

jtmcg commented on acceptance/testdata/auth/auth-switch.txtar:
```diff
@@ -0,0 +1,58 @@
+# This requires a second account to log in to, meaning the GH_USER, 
```
🤔 I'm wondering if there's a way to skip this test if any of the optional env vars aren't set... Or maybe we just skip it by default?

williammartin replied:
I'm noodling on an idea like:

```
requireenv
```

or

```
skip-if-missing-env
```

These would only be accepted by default but in CI we would have a `GH_ACCEPTANCE_STRICT=true` that doesn't allow this, so that tests aren't skipped but fail. Or we could just assert no skipped tests. Just different tradeoffs.

jtmcg replied:
🤔 Different tradeoffs, for sure... I'd probably want to see how many more Env Vars we're gonna need to get more of the full test suite implemented to understand how flexible it needs to be. Definitely out of scope of this

williammartin commented on acceptance/testdata/auth/auth-status.txtar:
```diff
@@ -0,0 +1,3 @@
+# Check the authentication status
```
**Food For Thought**

Since the host details are kept locally, there's nothing in this test that is really validating we're working with the remote correctly. Perhaps checking for `✓ Logged in to` or something might also be useful?

williammartin commented on acceptance/testdata/auth/auth-login-logout.txtar:
```diff
@@ -0,0 +1,31 @@
+# We aren't logged in at the moment, but GH_HOST will override the
```
**Change Requested**

I think you mean `GH_TOKEN` in this comment.

jtmcg replied:
I do. Good catch 👍

williammartin commented on acceptance/testdata/auth/auth-login-logout.txtar:
```diff
@@ -0,0 +1,31 @@
+# Copy $GH_TOKEN to a new env var
+env LOGIN_TOKEN=$GH_TOKEN
+exec echo $LOGIN_TOKEN
+stdout $GH_TOKEN
```
**Suggestion**

I think you were probably using these to validate your understanding of testscript behaviour but I think they detract from the test. I'd suggest you remove them.

williammartin commented on acceptance/testdata/auth/auth-login-logout.txtar:
```diff
@@ -0,0 +1,31 @@
+
+# Ensure the token was deleted
+exec echo $GH_TOKEN
+! stdout '.'
```
**Suggestion**

I feel the same way about this as the suggestion above. I don't think these assertions do much except convince you during development that the test is working the way you expect.

I think we should have confidence that once we have taught ourselves the behaviour of commands, that they work as expected. It doesn't really buy us any extra confidence because `auth login` will fail if the token is set.

williammartin commented on acceptance/testdata/auth/auth-login-logout.txtar:
```diff
@@ -0,0 +1,31 @@
+
+# Login to the host by feeding the token to stdin
+exec echo $LOGIN_TOKEN
+stdin stdout
```
**Interesting**

It's surprising to me that there's no way to pass a string directly to stdin! Interesting thing to know.

Unfortunately this is the first place we've taken a dependency on something other than `gh` and `git`. I'm not exactly sure what the state of these tests on windows is but we may want to create a custom command in future to avoid this. Not something to think too hard about for the moment though.

williammartin commented on acceptance/testdata/auth/auth-setup-git.txtar:
```diff
@@ -0,0 +1,13 @@
+# Setup the credential key name string
+formattedStringToEnv CREDENTIAL_HELPER_KEY 'credential.https://%s.helper' $GH_HOST
```
**Change Requested**

I think this is equivalent to:

```
env CREDENTIAL_HELPER_KEY=credential.https://${GH_HOST}.helper
```

So we can hopefully remove the command?

jtmcg replied:
Nice! It does work. Much cleaner. In fact, I've removed the `CREDENTIAL_HELPER_KEY` overall in favor of the string creation with the variable.

Note, I was having issues finding out how to do that in the docs. It's helpful to know that `${ENV_VAR}` works for building strings like this 👍

williammartin commented on acceptance/testdata/auth/auth-setup-git.txtar:
```diff
@@ -0,0 +1,13 @@
+
+# Check that the credential helper is set to gh
+exec git config --get $CREDENTIAL_HELPER_KEY
+stdout gh
```
**Suggestion**

```
stdout '^.*gh auth git-credential$'
```

I think this might give a bit more context to the assertion.

williammartin commented on acceptance/testdata/auth/auth-switch.txtar:
```diff
@@ -0,0 +1,58 @@
+stdin stdout
+exec gh auth login --hostname=$GH_HOST_ALT --with-token --insecure-storage
+
+# Confirm the active user is GH_USER_ALT
+exec gh auth status --active
+stdout $GH_USER_ALT
+
+# Confirm that the active host is GH_HOST_ALT
+exec gh auth status --active
+stdout $GH_HOST_ALT
+
+# Switch to the original user
+exec gh auth switch -h=$GH_HOST -u=$GH_USER
+
+# Confirm that the active user is GH_USER
+exec gh auth status --active
+stdout $GH_USER
+
+# Confirm that the active host is GH_HOST
+exec gh auth status --active
```
I don't think this test is doing what you think it is. There's no such thing as an `--active` host. Here's my (redacted) output:

```
➜ gh auth status --active
github.com
  ✓ Logged in to github.com account williammartin (keyring)
  - Active account: true
  - Git operations protocol: https
  - Token: gho_************************************
  - Token scopes: 'delete_repo', 'gist', 'read:org', 'repo', 'workflow'

<other>
  ✓ Logged in to <other> account williammartin (keyring)
  - Active account: true
  - Git operations protocol: https
  - Token: gho_************************************
  - Token scopes: 'delete_repo', 'gist', 'read:org', 'repo', 'workflow'
```

So each host has an active user. If `GH_HOST` and `GH_HOST_ALT` are the same, there will be one entry but it will always be the same host. If they are different hosts there will be two entries but they will both be there. So an assertion for an active host can never fail.

Similarly, imagine that `GH_HOST` and `GH_HOST_ALT` are different, then the assertion on the active user can never fail because each host only has one user, and therefore that user is active and will appear in the output.

jtmcg replied:
Interesting... I was tinkering around with this with a very specific login state, which _may_ have allowed the behavior here to assert what I wanted it to, but that doesn't mean that it will be for all configurations of Hosts and Users, which would definitely result in flakiness... Good callout!

I think the behavior I actually want is `gh auth status --active -h $GH_HOST` and I need to remove the`GH_HOST_ALT` env var, essentially forcing this to be tested against two different user accounts on the same host.

That said, I'm not sure how valuable that is... 🤔

williammartin replied:
Yes I think that's what I would expect. It's very valuable as a UX test for `auth switch` but a little less valuable for the GHEC data residency confidence since the command is entirely local.

williammartin commented on acceptance/acceptance_test.go:
```diff
@@ -150,6 +159,12 @@ func sharedSetup(tsEnv testScriptEnv) func(ts *testscript.Env) error {
 		ts.Setenv("GH_TOKEN", tsEnv.token)
 
+		// Used for gh auth testing
+		ts.Setenv("GH_USER", tsEnv.user)
```
**Food For Thought**

Don't need to get ahead of ourselves here but this looks like not _shared setup_, and maybe we need some other way to customise this for specific commands / testscripts.

jtmcg committed:
Address PR feedback

williammartin reviewed (approved):
```
➜  cli git:(jtmcg/testscripts-auth) set -o pipefail && GH_ACCEPTANCE_HOST=github.com GH_ACCEPTANCE_ORG=gh-acceptance-testing go test -tags acceptance -json -coverprofile=coverage.out -coverpkg=./... -run ^TestAuth$ github.com/cli/cli/v2/acceptance | tparse --all go test
┌────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │            TEST            │             PACKAGE               │
│─────────┼─────────┼────────────────────────────┼───────────────────────────────────│
│  PASS   │    1.67 │ TestAuth/auth-login-logout │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    1.12 │ TestAuth/auth-status       │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    0.72 │ TestAuth/auth-setup-git    │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    0.64 │ TestAuth/auth-token        │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    0.00 │ TestAuth                   │ github.com/cli/cli/v2/acceptance  │
└────────────────────────────────────────────────────────────────────────────────────┘
┌────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │             PACKAGE              │ COVER │ PASS │ FAIL │ SKIP  │
│─────────┼─────────┼──────────────────────────────────┼───────┼──────┼──────┼───────│
│  PASS   │  2.30s  │ github.com/cli/cli/v2/acceptance │ 9.1%  │  5   │  0   │  0    │
└────────────────────────────────────────────────────────────────────────────────────┘
```

LGTM!
